core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java

```diff
@@ -30,30 +32,77 @@ public class PathfindingResult {
     /**
      * pathfindingResultMake is used to create a class PathfindingResult with a good format to be sent to the middleware
      * during this function, we add geometric information in the PathfindingResult
+     *
      * @param path contains the pathfinding's result
      */
     public static PathfindingResult make(
-            List<Pathfinding.EdgeRange<SignalingRoute>> path,
+            Pathfinding.Result<SignalingRoute> path,
             SignalingInfra infra,
             WarningRecorderImpl warningRecorder
     ) {
         var res = new PathfindingResult();
-        for (var signalingRouteEdgeRange : path)
+        for (var signalingRouteEdgeRange : path.ranges())
             if (signalingRouteEdgeRange.start() < signalingRouteEdgeRange.end())
                 res.routePaths.add(
                         makeRouteResult(
                                 res,
                                 signalingRouteEdgeRange
                         )
                 );
-        var lastRoute = res.routePaths.get(res.routePaths.size() - 1);
-        var lastRange = lastRoute.trackSections.get(lastRoute.trackSections.size() - 1);
-        res.addStep(new PathWaypointResult(lastRange.trackSection.id.id, lastRange.end));
         res.addGeometry(infra);
         res.warnings = warningRecorder.warnings;
+        setAllWaypoints(res, path);
         return res;
     }
 
+    /** Adds all missing waypoints and sorts them */
+    private static void setAllWaypoints(PathfindingResult res, Pathfinding.Result<SignalingRoute> path) {
+        for (var waypoint : path.waypoints())
+            res.addStep(convertLocation(waypoint));
+        var waypointsPerTrack = HashMultimap.<String, PathWaypointResult>create();
+        for (var waypoint : res.pathWaypoints)
+            waypointsPerTrack.put(waypoint.track.id.id, waypoint);
+        double routeOffset = 0;
+        for (var routeRange : path.ranges()) {
+            double trackOffset = routeOffset - routeRange.start();
+            for (var trackRange : routeRange.edge().getInfraRoute().getTrackRanges()) {
+                for (var waypoint : waypointsPerTrack.get(trackRange.track.getEdge().getID())) {
+                    if (trackRange.containsOffset(waypoint.position)) {
+                        waypoint.pathOffset = trackOffset + trackRange.offsetOf(waypoint.position);
+                    }
+                }
+                trackOffset += trackRange.getLength();
+            }
+            routeOffset += routeRange.end() - routeRange.start();
+        }
```

Comment:
Here's an idea: iterate on all the waypoints in the path, and add them in order. When doing that, you can determine if the waypoint was added by the user or if it's a suggestion. this way, no need to compute the path offset or to sort, as the result if guaranteed to be sorted by the way we iterate over waypoints.